### ISSUE TITLE
Kafka channel webhook in crash loop

### ISSUE THREAD

**Issue by AceHack** [knative/eventing-contrib#1554]
**Describe the bug**
Kafka channel webhook crashes over and over with:
`2020/09/10 13:43:00 http: TLS handshake error from 10.78.180.106:33934: server key missing`
Also, the `messaging-webhook-certs` never gets updated with real certs

 **Expected behavior**
Webhook to startup correctly and certs created

 **To Reproduce**
Install kafka channel

 **Knative release version**
serving 0.17.2, eventing 0.17.3, kafka channel contrib 0.17.1

 **Additional context**
As a work around, If I add the following to the Kafka channel webhook it works

```
          args:
            - "--disable-ha"
```

**Comment by slinkydeveloper** [comment]
Weird, i wonder why disable-ha affects the webhooks. Let me try it

**Comment by slinkydeveloper** [comment]
I tried to reproduce the issue, installing on kind using these scripts https://github.com/matzew/kn-box and the webhook certs secret is properly populated:

```
(master) % kubectl describe secrets -n knative-eventing messaging-webhook-certs
Name:         messaging-webhook-certs
Namespace:    knative-eventing
Labels:       contrib.eventing.knative.dev/release=v0.17.0
Annotations:  <none>

Type:  Opaque

Data
====
ca-cert.pem:      1415 bytes
server-cert.pem:  1399 bytes
server-key.pem:   1679 bytes
```

**Comment by AceHack** [comment]
I was installing with 0.17.1, not 0.17.0.  Looking for other differences in the script.

**Comment by AceHack** [comment]
Also not sure if it makes any difference but I'm using istio not kourier and I'm using the operator for installing serving and eventing.

**Comment by slinkydeveloper** [comment]
Maybe that's an issue in the operator @aliok ?

**Comment by AceHack** [comment]
@slinkydeveloper The script you point to is also doing something weird with the sugar controller.

kubectl apply --filename https://storage.googleapis.com/knative-releases/eventing/previous/${eventing_version}/eventing-sugar-controller.yaml

**Comment by AceHack** [comment]
@slinkydeveloper I wrote this script and I am unable to reproduce this issue locally in docker desktop currently.  Gonna update it to use the operators and see if that makes a difference.  Also looking for any differences to our dev environment.

```
#!/usr/bin/env bash

function header_text {
  echo "$header$*$reset"
}

CONTEXT=docker-desktop
SERVING=0.17.2
EVENTING=0.17.3
EVENTING_CONTRIB=0.17.1
ISTIO=$(curl https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/istio-stable)
STRIMZI=`curl https://github.com/strimzi/strimzi-kafka-operator/releases/latest |  awk -F 'tag/' '{print $2}' | awk -F '"' '{print $1}' 2>/dev/null`

header_text "Setup context"
kubectl config use-context ${CONTEXT}

header_text "Creating namespaces"
kubectl create ns istio-system
kubectl create ns cert-manager
kubectl create ns knative-serving
kubectl create ns knative-eventing
kubectl create namespace kafka


header_text "Setting up Istio"
kubectl apply -f "https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/${ISTIO}/istio-crds.yaml"
while [[ $(kubectl get crd gateways.networking.istio.io -o jsonpath='{.status.conditions[?(@.type=="Established")].status}') != 'True' ]]; do
  echo "Waiting on Istio CRDs"; sleep 1
done
kubectl apply -f "https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/${ISTIO}/istio-minimal.yaml"

header_text "Waiting for Istio to become ready"
sleep 5; while echo && kubectl get pods -n istio-system | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Cert Manager"
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/cert-manager-0.12.0/cert-manager-crds.yaml"
while [[ $(kubectl get crd certificates.cert-manager.io -o jsonpath='{.status.conditions[?(@.type=="Established")].status}') != 'True' ]]; do
  echo "Waiting on Cert-Manager CRDs"; sleep 1
done
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/cert-manager-0.12.0/cert-manager.yaml"

header_text "Waiting for Cert Manager to become ready"
sleep 5; while echo && kubectl get pods -n cert-manager | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Serving"
kubectl apply --filename https://github.com/knative/serving/releases/download/v${SERVING}/serving-crds.yaml
 n=0
   until [ $n -ge 2 ]
   do
      kubectl apply --filename https://github.com/knative/serving/releases/download/v${SERVING}/serving-core.yaml && break
      n=$[$n+1]
      sleep 5
   done
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/net-istio.yaml"

header_text "Waiting for Serving to become ready"
sleep 5; while echo && kubectl get pods -n knative-serving | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Eventing"
kubectl apply --filename https://github.com/knative/eventing/releases/download/v${EVENTING}/eventing-crds.yaml
kubectl apply --filename https://github.com/knative/eventing/releases/download/v${EVENTING}/eventing-core.yaml
kubectl apply --filename https://github.com/knative/eventing/releases/download/v${EVENTING}/in-memory-channel.yaml

header_text "Waiting for Eventing to become ready"
sleep 5; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Strimzi install"
kubectl -n kafka apply --selector strimzi.io/crd-install=true -f https://github.com/strimzi/strimzi-kafka-operator/releases/download/${STRIMZI}/strimzi-cluster-operator-${STRIMZI}.yaml
curl -L "https://github.com/strimzi/strimzi-kafka-operator/releases/download/${STRIMZI}/strimzi-cluster-operator-${STRIMZI}.yaml" \
  | sed 's/namespace: .*/namespace: kafka/' \
  | kubectl -n kafka apply -f -

header_text "Applying Strimzi Cluster file"
kubectl -n kafka apply -f "https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/${STRIMZI}/examples/kafka/kafka-persistent-single.yaml"

header_text "Waiting for Strimzi to become ready"
sleep 5; while echo && kubectl get pods -n kafka | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Knative Apache Kafka Channel"
curl -L "https://github.com/knative/eventing-contrib/releases/download/v${EVENTING_CONTRIB}/kafka-channel.yaml" \
    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
    | kubectl apply --filename -

header_text "Waiting for Knative Apache Kafka Channel to become ready"
sleep 5; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done
```

**Comment by AceHack** [comment]
@slinkydeveloper 
Okay, I am able to reproduce the issue now locally.  Run the above script that creates everything.  After it finishes create a topic with.

```
cat <<-EOF | kubectl apply -f -
---
apiVersion: messaging.knative.dev/v1beta1
kind: KafkaChannel
metadata:
  name: my-kafka-channel
spec:
  numPartitions: 3
  replicationFactor: 1
EOF
```

Now run
```
kubectl delete deployment -n knative-eventing kafka-ch-controller
kubectl delete deployment -n knative-eventing kafka-webhook
kubectl delete secret -n knative-eventing messaging-webhook-certs
```

Re-apply the kafka channel with
```
curl -L "https://github.com/knative/eventing-contrib/releases/download/v${EVENTING_CONTRIB}/kafka-channel.yaml" \
    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
    | kubectl apply --filename -
```

Now you will see the issue.

```
2020/09/14 18:38:29 maxprocs: Leaving GOMAXPROCS=10: CPU quota undefined
2020/09/14 18:38:29 Registering 2 clients
2020/09/14 18:38:29 Registering 3 informer factories
2020/09/14 18:38:29 Registering 4 informers
2020/09/14 18:38:29 Registering 4 controllers
{"level":"info","ts":"2020-09-14T18:38:29.935Z","caller":"logging/config.go:110","msg":"Successfully created the logger."}
{"level":"info","ts":"2020-09-14T18:38:29.935Z","caller":"logging/config.go:111","msg":"Logging level set to info"}
{"level":"info","ts":"2020-09-14T18:38:29.935Z","caller":"logging/config.go:78","msg":"Fetch GitHub commit ID from kodata failed","error":"open /var/run/ko/HEAD: no such file or directory"}
{"level":"info","ts":"2020-09-14T18:38:29.937Z","logger":"kafkachannel-webhook","caller":"profiling/server.go:59","msg":"Profiling enabled: false"}
{"level":"info","ts":"2020-09-14T18:38:29.951Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:46","msg":"Running with Standard leader election"}
{"level":"info","ts":"2020-09-14T18:38:29.969Z","logger":"kafkachannel-webhook","caller":"sharedmain/main.go:228","msg":"Starting configuration manager..."}
{"level":"info","ts":"2020-09-14T18:38:29.981Z","logger":"kafkachannel-webhook","caller":"metrics/exporter.go:152","msg":"Flushing the existing exporter before setting up the new exporter."}
{"level":"info","ts":"2020-09-14T18:38:29.981Z","logger":"kafkachannel-webhook","caller":"metrics/prometheus_exporter.go:50","msg":"Created Opencensus Prometheus exporter with config: &{knative.dev/eventing kafkachannel_webhook prometheus 5000000000 <nil> <nil>  false 9090 false   {   false}}. Start the server for Prometheus exporter."}
{"level":"info","ts":"2020-09-14T18:38:29.981Z","logger":"kafkachannel-webhook","caller":"metrics/exporter.go:165","msg":"Successfully updated the metrics exporter; old config: <nil>; new config &{knative.dev/eventing kafkachannel_webhook prometheus 5000000000 <nil> <nil>  false 9090 false   {   false}}"}
{"level":"info","ts":"2020-09-14T18:38:30.074Z","logger":"kafkachannel-webhook","caller":"sharedmain/main.go:249","msg":"Starting informers..."}
2020/09/14 18:38:30 http: TLS handshake error from 10.1.0.1:50700: server key missing
{"level":"info","ts":"2020-09-14T18:38:30.574Z","logger":"kafkachannel-webhook","caller":"webhook/webhook.go:169","msg":"Informers have been synced, unblocking admission webhooks."}
{"level":"info","ts":"2020-09-14T18:38:30.574Z","logger":"kafkachannel-webhook","caller":"sharedmain/main.go:256","msg":"Starting controllers..."}
{"level":"info","ts":"2020-09-14T18:38:30.575Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:134","msg":"kafkachannel-webhook.conversionwebhook.00-of-01 will run in leader-elected mode with id \"kafka-webhook-7796d54ddf-p8xjk_77256b20-fba5-46b8-ae8a-e31a93411053\""}
{"level":"info","ts":"2020-09-14T18:38:30.575Z","logger":"kafkachannel-webhook.ConversionWebhook","caller":"controller/controller.go:444","msg":"Starting controller and workers"}
{"level":"info","ts":"2020-09-14T18:38:30.575Z","logger":"kafkachannel-webhook.ConversionWebhook","caller":"controller/controller.go:454","msg":"Started workers"}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:134","msg":"kafkachannel-webhook.defaultingwebhook.00-of-01 will run in leader-elected mode with id \"kafka-webhook-7796d54ddf-p8xjk_eeb0f9d2-9166-431f-b9a1-da33e639bac1\""}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook.DefaultingWebhook","caller":"controller/controller.go:444","msg":"Starting controller and workers"}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook.ConversionWebhook","caller":"controller/controller.go:520","msg":"Reconcile succeeded. Time taken: 54.2µs","knative.dev/traceid":"4d35f399-5434-4781-a99c-c4349e6880f2","knative.dev/key":"kafkachannels.messaging.knative.dev"}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook.DefaultingWebhook","caller":"controller/controller.go:454","msg":"Started workers"}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:134","msg":"kafkachannel-webhook.validationwebhook.00-of-01 will run in leader-elected mode with id \"kafka-webhook-7796d54ddf-p8xjk_43a4f525-b6a0-477a-aa97-c2f18cb66db0\""}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook.DefaultingWebhook","caller":"controller/controller.go:520","msg":"Reconcile succeeded. Time taken: 42µs","knative.dev/traceid":"8eb870c7-d9b8-44f2-b52b-31e8d16cd867","knative.dev/key":"knative-eventing/messaging-webhook-certs"}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook.ValidationWebhook","caller":"controller/controller.go:444","msg":"Starting controller and workers"}
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook.ValidationWebhook","caller":"controller/controller.go:454","msg":"Started workers"}
I0914 18:38:30.577114       1 leaderelection.go:242] attempting to acquire leader lease  knative-eventing/kafkachannel-webhook.conversionwebhook.00-of-01...
{"level":"info","ts":"2020-09-14T18:38:30.576Z","logger":"kafkachannel-webhook.DefaultingWebhook","caller":"controller/controller.go:520","msg":"Reconcile succeeded. Time taken: 53.4µs","knative.dev/traceid":"83d34258-d069-4b5c-baef-223cf9e1b461","knative.dev/key":"defaulting.webhook.kafka.messaging.knative.dev"}
{"level":"info","ts":"2020-09-14T18:38:30.577Z","logger":"kafkachannel-webhook.ValidationWebhook","caller":"controller/controller.go:520","msg":"Reconcile succeeded. Time taken: 41.5µs","knative.dev/traceid":"87728b65-6b25-40e1-acd2-60061e3fd6c1","knative.dev/key":"knative-eventing/messaging-webhook-certs"}
{"level":"info","ts":"2020-09-14T18:38:30.577Z","logger":"kafkachannel-webhook.ValidationWebhook","caller":"controller/controller.go:520","msg":"Reconcile succeeded. Time taken: 40.2µs","knative.dev/traceid":"18dcce3a-fe6e-420b-a7b0-a74096161d0e","knative.dev/key":"validation.webhook.kafka.messaging.knative.dev"}
I0914 18:38:30.577935       1 leaderelection.go:242] attempting to acquire leader lease  knative-eventing/kafkachannel-webhook.validationwebhook.00-of-01...
{"level":"info","ts":"2020-09-14T18:38:30.577Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:134","msg":"kafkachannel-webhook.webhookcertificates.00-of-01 will run in leader-elected mode with id \"kafka-webhook-7796d54ddf-p8xjk_51c2f091-9486-4dbc-a147-31bf6bf99c6b\""}
{"level":"info","ts":"2020-09-14T18:38:30.578Z","logger":"kafkachannel-webhook.WebhookCertificates","caller":"controller/controller.go:444","msg":"Starting controller and workers"}
{"level":"info","ts":"2020-09-14T18:38:30.578Z","logger":"kafkachannel-webhook.WebhookCertificates","caller":"controller/controller.go:454","msg":"Started workers"}
{"level":"info","ts":"2020-09-14T18:38:30.578Z","logger":"kafkachannel-webhook.WebhookCertificates","caller":"controller/controller.go:520","msg":"Reconcile succeeded. Time taken: 37.1µs","knative.dev/traceid":"824c7d44-d815-4473-b3e8-9b68ca99b9b8","knative.dev/key":"knative-eventing/messaging-webhook-certs"}
I0914 18:38:30.578749       1 leaderelection.go:242] attempting to acquire leader lease  knative-eventing/kafkachannel-webhook.defaultingwebhook.00-of-01...
I0914 18:38:30.579217       1 leaderelection.go:242] attempting to acquire leader lease  knative-eventing/kafkachannel-webhook.webhookcertificates.00-of-01...
2020/09/14 18:38:30 http: TLS handshake error from 10.1.0.1:50722: server key missing
2020/09/14 18:38:31 http: TLS handshake error from 10.1.0.1:50724: server key missing
2020/09/14 18:38:31 http: TLS handshake error from 10.1.0.1:50746: server key missing
2020/09/14 18:38:32 http: TLS handshake error from 10.1.0.1:50752: server key missing
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:152","msg":"\"kafka-webhook-7796d54ddf-p8xjk_51c2f091-9486-4dbc-a147-31bf6bf99c6b\" has stopped leading \"kafkachannel-webhook.webhookcertificates.00-of-01\""}
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:152","msg":"\"kafka-webhook-7796d54ddf-p8xjk_43a4f525-b6a0-477a-aa97-c2f18cb66db0\" has stopped leading \"kafkachannel-webhook.validationwebhook.00-of-01\""}
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:152","msg":"\"kafka-webhook-7796d54ddf-p8xjk_eeb0f9d2-9166-431f-b9a1-da33e639bac1\" has stopped leading \"kafkachannel-webhook.defaultingwebhook.00-of-01\""}
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook.WebhookCertificates","caller":"controller/controller.go:456","msg":"Shutting down workers"}
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook","caller":"leaderelection/context.go:152","msg":"\"kafka-webhook-7796d54ddf-p8xjk_77256b20-fba5-46b8-ae8a-e31a93411053\" has stopped leading \"kafkachannel-webhook.conversionwebhook.00-of-01\""}
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook.ValidationWebhook","caller":"controller/controller.go:456","msg":"Shutting down workers"}
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook.DefaultingWebhook","caller":"controller/controller.go:456","msg":"Shutting down workers"}
{"level":"info","ts":"2020-09-14T18:38:32.226Z","logger":"kafkachannel-webhook.ConversionWebhook","caller":"controller/controller.go:456","msg":"Shutting down workers"}
{"level":"info","ts":"2020-09-14T18:38:32.227Z","logger":"kafkachannel-webhook","caller":"webhook/webhook.go:225","msg":"Starting to fail readiness probes..."}
2020/09/14 18:38:32 http: TLS handshake error from 10.1.0.1:50770: server key missing
2020/09/14 18:38:33 http: TLS handshake error from 10.1.0.1:50790: server key missing
2020/09/14 18:38:34 http: TLS handshake error from 10.1.0.1:50818: server key missing
2020/09/14 18:38:35 http: TLS handshake error from 10.1.0.1:50850: server key missing
2020/09/14 18:38:36 http: TLS handshake error from 10.1.0.1:50874: server key missing
2020/09/14 18:38:37 http: TLS handshake error from 10.1.0.1:50892: server key missing
2020/09/14 18:38:38 http: TLS handshake error from 10.1.0.1:50922: server key missing
2020/09/14 18:38:39 http: TLS handshake error from 10.1.0.1:50942: server key missing
2020/09/14 18:38:40 http: TLS handshake error from 10.1.0.1:50970: server key missing
2020/09/14 18:38:41 http: TLS handshake error from 10.1.0.1:50992: server key missing
2020/09/14 18:38:42 http: TLS handshake error from 10.1.0.1:51014: server key missing
2020/09/14 18:38:43 http: TLS handshake error from 10.1.0.1:51034: server key missing
2020/09/14 18:38:44 http: TLS handshake error from 10.1.0.1:51060: server key missing
2020/09/14 18:38:45 http: TLS handshake error from 10.1.0.1:51088: server key missing
2020/09/14 18:38:46 http: TLS handshake error from 10.1.0.1:51112: server key missing
2020/09/14 18:38:47 http: TLS handshake error from 10.1.0.1:51128: server key missing
2020/09/14 18:38:48 http: TLS handshake error from 10.1.0.1:51156: server key missing
2020/09/14 18:38:49 http: TLS handshake error from 10.1.0.1:51176: server key missing
2020/09/14 18:38:50 http: TLS handshake error from 10.1.0.1:51204: server key missing
2020/09/14 18:38:51 http: TLS handshake error from 10.1.0.1:51226: server key missing
2020/09/14 18:38:52 http: TLS handshake error from 10.1.0.1:51248: server key missing
2020/09/14 18:38:53 http: TLS handshake error from 10.1.0.1:51268: server key missing
2020/09/14 18:38:54 http: TLS handshake error from 10.1.0.1:51294: server key missing
2020/09/14 18:38:55 http: TLS handshake error from 10.1.0.1:51322: server key missing
2020/09/14 18:38:56 http: TLS handshake error from 10.1.0.1:51346: server key missing
2020/09/14 18:38:57 http: TLS handshake error from 10.1.0.1:51362: server key missing
2020/09/14 18:38:58 http: TLS handshake error from 10.1.0.1:51390: server key missing
2020/09/14 18:38:59 http: TLS handshake error from 10.1.0.1:51410: server key missing
2020/09/14 18:39:00 http: TLS handshake error from 10.1.0.1:51438: server key missing
2020/09/14 18:39:01 http: TLS handshake error from 10.1.0.1:51460: server key missing
2020/09/14 18:39:02 http: TLS handshake error from 10.1.0.1:51482: server key missing
2020/09/14 18:39:03 http: TLS handshake error from 10.1.0.1:51502: server key missing
2020/09/14 18:39:04 http: TLS handshake error from 10.1.0.1:51530: server key missing
2020/09/14 18:39:05 http: TLS handshake error from 10.1.0.1:51558: server key missing
2020/09/14 18:39:06 http: TLS handshake error from 10.1.0.1:51582: server key missing
2020/09/14 18:39:07 http: TLS handshake error from 10.1.0.1:51600: server key missing
2020/09/14 18:39:08 http: TLS handshake error from 10.1.0.1:51630: server key missing
2020/09/14 18:39:09 http: TLS handshake error from 10.1.0.1:51650: server key missing
2020/09/14 18:39:10 http: TLS handshake error from 10.1.0.1:51680: server key missing
2020/09/14 18:39:11 http: TLS handshake error from 10.1.0.1:51702: server key missing
2020/09/14 18:39:12 http: TLS handshake error from 10.1.0.1:51724: server key missing
2020/09/14 18:39:13 http: TLS handshake error from 10.1.0.1:51744: server key missing
2020/09/14 18:39:14 http: TLS handshake error from 10.1.0.1:51768: server key missing
2020/09/14 18:39:15 http: TLS handshake error from 10.1.0.1:51798: server key missing
2020/09/14 18:39:16 http: TLS handshake error from 10.1.0.1:51820: server key missing
```

**Comment by AceHack** [comment]
FYI, in case you are interested here is the script I created to create everything the operator way.  The issue can be reproduced both with and without the operator.

```
#!/usr/bin/env bash

function header_text {
  echo "$header$*$reset"
}

CONTEXT=docker-desktop
OPERATOR=0.17.1
SERVING=0.17.2
EVENTING_CONTRIB=0.17.1
ISTIO=$(curl https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/istio-stable)
STRIMZI=`curl https://github.com/strimzi/strimzi-kafka-operator/releases/latest |  awk -F 'tag/' '{print $2}' | awk -F '"' '{print $1}' 2>/dev/null`

header_text "Setup context"
kubectl config use-context ${CONTEXT}

header_text "Creating namespaces"
kubectl create ns istio-system
kubectl create ns cert-manager
kubectl create ns knative-serving
kubectl create ns knative-eventing
kubectl create namespace kafka


header_text "Setting up Istio"
kubectl apply -f "https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/${ISTIO}/istio-crds.yaml"
while [[ $(kubectl get crd gateways.networking.istio.io -o jsonpath='{.status.conditions[?(@.type=="Established")].status}') != 'True' ]]; do
  echo "Waiting on Istio CRDs"; sleep 1
done
kubectl apply -f "https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/${ISTIO}/istio-minimal.yaml"

header_text "Waiting for Istio to become ready"
sleep 5; while echo && kubectl get pods -n istio-system | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Cert Manager"
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/cert-manager-0.12.0/cert-manager-crds.yaml"
while [[ $(kubectl get crd certificates.cert-manager.io -o jsonpath='{.status.conditions[?(@.type=="Established")].status}') != 'True' ]]; do
  echo "Waiting on Cert-Manager CRDs"; sleep 1
done
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/cert-manager-0.12.0/cert-manager.yaml"

header_text "Waiting for Cert Manager to become ready"
sleep 5; while echo && kubectl get pods -n cert-manager | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Operator"
kubectl apply --filename https://github.com/knative/operator/releases/download/v${OPERATOR}/operator.yaml

header_text "Waiting for Operator to become ready"
sleep 5; while echo && kubectl get pods -n default | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done; sleep 10;


header_text "Setting up Serving"
cat <<-EOF | kubectl apply -f -
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
EOF

header_text "Waiting for Serving to become ready"
sleep 20; while echo && kubectl get pods -n knative-serving | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Eventing"
cat <<-EOF | kubectl apply -f -
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
EOF

header_text "Waiting for Eventing to become ready"
sleep 20; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Strimzi install"
kubectl -n kafka apply --selector strimzi.io/crd-install=true -f https://github.com/strimzi/strimzi-kafka-operator/releases/download/${STRIMZI}/strimzi-cluster-operator-${STRIMZI}.yaml
curl -L "https://github.com/strimzi/strimzi-kafka-operator/releases/download/${STRIMZI}/strimzi-cluster-operator-${STRIMZI}.yaml" \
  | sed 's/namespace: .*/namespace: kafka/' \
  | kubectl -n kafka apply -f -

header_text "Applying Strimzi Cluster file"
kubectl -n kafka apply -f "https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/${STRIMZI}/examples/kafka/kafka-persistent-single.yaml"

header_text "Waiting for Strimzi to become ready"
sleep 5; while echo && kubectl get pods -n kafka | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Knative Apache Kafka Channel"
curl -L "https://github.com/knative/eventing-contrib/releases/download/v${EVENTING_CONTRIB}/kafka-channel.yaml" \
    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
    | kubectl apply --filename -

header_text "Waiting for Knative Apache Kafka Channel to become ready"
sleep 5; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done
```

**Comment by AceHack** [comment]
@slinkydeveloper Here is one script to reproduce the whole thing, no operator involved.

```
#!/usr/bin/env bash

function header_text {
  echo "$header$*$reset"
}

CONTEXT=docker-desktop
SERVING=0.17.2
EVENTING=0.17.3
EVENTING_CONTRIB=0.17.1
ISTIO=$(curl https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/istio-stable)
STRIMZI=`curl https://github.com/strimzi/strimzi-kafka-operator/releases/latest |  awk -F 'tag/' '{print $2}' | awk -F '"' '{print $1}' 2>/dev/null`

header_text "Setup context"
kubectl config use-context ${CONTEXT}

header_text "Creating namespaces"
kubectl create ns istio-system
kubectl create ns cert-manager
kubectl create ns knative-serving
kubectl create ns knative-eventing
kubectl create namespace kafka


header_text "Setting up Istio"
kubectl apply -f "https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/${ISTIO}/istio-crds.yaml"
while [[ $(kubectl get crd gateways.networking.istio.io -o jsonpath='{.status.conditions[?(@.type=="Established")].status}') != 'True' ]]; do
  echo "Waiting on Istio CRDs"; sleep 1
done
kubectl apply -f "https://raw.githubusercontent.com/knative-sandbox/net-istio/master/third_party/${ISTIO}/istio-minimal.yaml"

header_text "Waiting for Istio to become ready"
sleep 5; while echo && kubectl get pods -n istio-system | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Cert Manager"
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/cert-manager-0.12.0/cert-manager-crds.yaml"
while [[ $(kubectl get crd certificates.cert-manager.io -o jsonpath='{.status.conditions[?(@.type=="Established")].status}') != 'True' ]]; do
  echo "Waiting on Cert-Manager CRDs"; sleep 1
done
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/cert-manager-0.12.0/cert-manager.yaml"

header_text "Waiting for Cert Manager to become ready"
sleep 5; while echo && kubectl get pods -n cert-manager | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Serving"
kubectl apply --filename https://github.com/knative/serving/releases/download/v${SERVING}/serving-crds.yaml
 n=0
   until [ $n -ge 2 ]
   do
      kubectl apply --filename https://github.com/knative/serving/releases/download/v${SERVING}/serving-core.yaml && break
      n=$[$n+1]
      sleep 5
   done
kubectl apply -f "https://github.com/knative/serving/raw/v${SERVING}/third_party/net-istio.yaml"

header_text "Waiting for Serving to become ready"
sleep 5; while echo && kubectl get pods -n knative-serving | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Eventing"
kubectl apply --filename https://github.com/knative/eventing/releases/download/v${EVENTING}/eventing-crds.yaml
kubectl apply --filename https://github.com/knative/eventing/releases/download/v${EVENTING}/eventing-core.yaml
kubectl apply --filename https://github.com/knative/eventing/releases/download/v${EVENTING}/in-memory-channel.yaml

header_text "Waiting for Eventing to become ready"
sleep 5; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Strimzi install"
kubectl -n kafka apply --selector strimzi.io/crd-install=true -f https://github.com/strimzi/strimzi-kafka-operator/releases/download/${STRIMZI}/strimzi-cluster-operator-${STRIMZI}.yaml
curl -L "https://github.com/strimzi/strimzi-kafka-operator/releases/download/${STRIMZI}/strimzi-cluster-operator-${STRIMZI}.yaml" \
  | sed 's/namespace: .*/namespace: kafka/' \
  | kubectl -n kafka apply -f -

header_text "Applying Strimzi Cluster file"
kubectl -n kafka apply -f "https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/${STRIMZI}/examples/kafka/kafka-persistent-single.yaml"

header_text "Waiting for Strimzi to become ready"
kubectl wait kafka -n kafka --all --for=condition="Ready"
sleep 5; while echo && kubectl get pods -n kafka | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Knative Apache Kafka Channel"
curl -L "https://github.com/knative/eventing-contrib/releases/download/v${EVENTING_CONTRIB}/kafka-channel.yaml" \
    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
    | kubectl apply --filename -

header_text "Waiting for Knative Apache Kafka Channel to become ready"
sleep 5; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Waiting for Kafka to become ready"
sleep 20
kubectl wait pods -n kafka --all --for=condition="Ready"


header_text "Creating kafka channel"
cat <<-EOF | kubectl apply -f -
---
apiVersion: messaging.knative.dev/v1beta1
kind: KafkaChannel
metadata:
  name: my-kafka-channel
spec:
  numPartitions: 3
  replicationFactor: 1
EOF
kubectl wait kafkachannel -n default --all --for=condition="Ready"


header_text "Deleting Knative Apache Kafka Channel"
kubectl delete --filename https://github.com/knative/eventing-contrib/releases/download/v${EVENTING_CONTRIB}/kafka-channel.yaml

header_text "Waiting for Knative Apache Kafka Channel to become deleted"
sleep 5; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done


header_text "Setting up Knative Apache Kafka Channel"
curl -L "https://github.com/knative/eventing-contrib/releases/download/v${EVENTING_CONTRIB}/kafka-channel.yaml" \
    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
    | kubectl apply --filename -

header_text "Waiting for Knative Apache Kafka Channel to become ready"
sleep 5; while echo && kubectl get pods -n knative-eventing | grep -v -E "(Running|Completed|STATUS)"; do sleep 5; done
```

**Comment by slinkydeveloper** [comment]
Yep I reproduced it. I've found another ticket in pkg that describes the same issue https://github.com/knative/eventing-contrib/issues/1554

**Comment by mattmoor** [comment]
If one of y'all have a repro setup, can you try changing the liveness probe to have one or both of:
 * `failureThreshold: 10`
 * `initialDelaySeconds: 10`

Judging from the logs, it is expecting 3 liveness probes to succeed over HTTPS unreasonably quickly after starting up.  The "has stopped leading" messages seem indicative of when the Pod is getting SIGTERM and the context is cancelled, which probably paints the picture of just how quickly we're expecting success 😞

**Comment by aliok** [comment]
For reference, https://github.com/knative-sandbox/eventing-kafka/pull/81 fixes this.
Backport for 0.18: https://github.com/knative/eventing-contrib/pull/1630
Backport for 0.17: https://github.com/knative/eventing-contrib/pull/1631

Won't be fixed on eventing-contrib master as we're not touching there anymore I think.

**Comment by aliok** [comment]
/close
This is fixed now with the PRs in comment above. There will be a new release for 0.17 and 0.18 next week.
